x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessHeaders.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.xpack.core.security.authc.RemoteAccessAuthentication;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public final class RemoteAccessHeaders {
+
+    public static final String REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY = "_remote_access_cluster_credential";
+    private final String clusterCredentialsHeader;
+    private final RemoteAccessAuthentication remoteAccessAuthentication;
+
+    public RemoteAccessHeaders(String clusterCredentialsHeader, RemoteAccessAuthentication remoteAccessAuthentication) {
+        assert clusterCredentialsHeader.startsWith("ApiKey ") : "cluster credentials headers must start with [ApiKey ]";
+        this.clusterCredentialsHeader = clusterCredentialsHeader;
+        this.remoteAccessAuthentication = remoteAccessAuthentication;
+    }
+
+    public void writeToContext(final ThreadContext ctx) throws IOException {
+        ctx.putHeader(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY, clusterCredentialsHeader);
+        remoteAccessAuthentication.writeToContext(ctx);
+    }
+
+    public static RemoteAccessHeaders readFromContext(final ThreadContext ctx) throws IOException {
+        final String clusterCredentialsHeader = ctx.getHeader(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY);
+        if (clusterCredentialsHeader == null) {
+            throw new IllegalArgumentException("remote access header [" + REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY + "] is required");
+        }
+        // Invoke parsing logic to validate that the header decodes to a valid API key credential
+        try (var ignored = parseClusterCredentialsHeader(clusterCredentialsHeader)) {} catch (Exception ex) {
+            throw new IllegalArgumentException(
+                "remote access header [" + REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY + "] value must be a valid API key credential",
+                ex
+            );
```

Comment:
Can this error throwing be moved inside `parseClusterCredentialsHeader` so that we just need to call `parseClusterCredentialsHeader(...)` here?

IIUC, the idea you are promoting with validation first before creation (or validation at creation time) is to avoid doing any work if credentials is not valid at all. This is generally a good security practice. I wonder whether we could push it a bit further: Instead of deserializing `RemoteAccessAuthentication` here, should we only store its encoded format and deserialize it after authentication with the credential proves to be successful? I don't think it needs to be part of this PR anyway. Mostly food for thought.

Reply:
> Can this error throwing be moved inside `parseClusterCredentialsHeader`

Yup good call. 

> Instead of deserializing `RemoteAccessAuthentication` here, should we only store its encoded format and deserialize it after authentication with the credential proves to be successful?

This might be a good idea, you're right. I'll consider it as part of the work around validating the authentication instance the querying cluster sends over. Seems like a natural time to address this. 

My primary motivation for force-parsing the API key above was mostly to avoid the un-intuitive behavior of calling what looks like a simple getter method leading to an exception, when it feels like if we managed to read `RemoteAccessHeader` from the thread context, it should be a valid object, ready for use (or fail on creation).

I think there's a balance to strike between how much of the data we should interpret before vs. after authc. We might need some info from `RemoteAccessAuthentication` for auditing (or logging) _before_ we've authenticated the API key, but either way I'll consider this in subsequent work.